### ISSUE TITLE
Do not allow using local variables in types

### ISSUE THREAD

**Issue by certik** [lcompilers/lpython#1554]
This should not be allowed:
```
def fill_A(k: i32, n: i32, b: CPtr) -> None:
    nk: i32 = n * k
    A: Pointer[i16[nk]] = c_p_pointer(b, i16[k * n])
```
The `nk` is a local variable, so it can't be part of a type.

**Comment by rebcabin** [comment]
Is the reason that the _value_ of a local variable might not be determinable until run time? If so, why not have (as a thought experiment) an option for generating code for run-time type checking? If not, what is the reason that local variables must not be allowed in dependent types?

**Comment by czgdp1807** [comment]
> Is the reason that the value of a local variable might not be determinable until run time?

We can allow variables in type definitions. However as you said we can only use the initial value of that variable, such as,

```python
nk: i32 = n * k
nk = n * k + 1
A: Pointer[i16[nk]] = c_p_pointer(b, i16[k * n])
```

Then, we would be using `n*k` for `nk` and not `n*k + 1`.



> If so, why not have (as a thought experiment) an option for generating code for run-time type checking?

As far as I can figure out from this, I think this is same as dynamic typing in CPython. So we won't be able to acheive the performance. Right now LPython is strictly typed so any discrepancy in types we raise errors (if we aren't then we should), so all kinds of optimisations are possible in such strictly typed code. However runtime type checking will give us a significant overhead. **Please correct me if your intent regarding run-time type checking is something else.**

Or if by run-time type checking you mean only checking the size of arrays at runtime, then that makes and shouldn't be causing much overhead. Like, `i16[size1]` and `i16[size2]` only differ in their sizes, rest is exactly same. So this is doable I feel. Still I personally would keep this for the distant future.

Let me know which one of the above you meant by runtime type checking.

> If not, what is the reason that local variables must not be allowed in dependent types?

We can allow, but as I said above then only the initial value of such variable will be used. Updates won't get reflected in the type.

**Comment by certik** [comment]
I don't think we want to do any kind of runtime type checking, for now we should only do strictly ahead of time compile time checking. So we can't use local variables at all. We can use parameters, like `nk: Const[i32] = 5`, but using function arguments in conctant parameters like this is also problematic (not allowed in Fortran for example), so I would not allow that either.

**Comment by rebcabin** [comment]
Sounds good. I'll post an Issue for opt-in run-time type checking in development / debug mode. Lots of compilers have something like that. We can consider this issue later.
